lib/cli/src/commands/create_exe.rs

```diff
@@ -344,1043 +256,1624 @@ impl CreateExe {
+        )
+        .to_string()));
+    }
 
-        write!(
-            c_code_to_instantiate,
-            "wasm_module_t *module = atom_{atom_to_run};"
+    // remove file if it exists - if not done, can lead to errors on copy
+    let _ = std::fs::remove_file(&normalize_path(&format!("{}", output_path.display())));
+    std::fs::copy(
+        &normalize_path(&format!("{}", out_path.display())),
+        &normalize_path(&format!("{}", output_path.display())),
+    )
+    .map_err(|e| {
+        anyhow::anyhow!(
+            "could not copy from {} to {}: {e}",
+            normalize_path(&format!("{}", out_path.display())),
+            normalize_path(&format!("{}", output_path.display()))
         )
-        .unwrap();
+    })?;
+
+    Ok(())
+}
 
-        WASMER_STATIC_MAIN_C_SOURCE
-            .replace("#define WASI", "#define WASI\r\n#define WASI_PIRITA")
-            .replace("// INSTANTIATE_MODULES", &c_code_to_instantiate)
-            .replace("##atom-name##", &atom_to_run)
-            .replace("wasm_module_delete(module);", &deallocate_module)
+pub(crate) fn normalize_path(s: &str) -> String {
+    if s.starts_with("\\\\?\\") {
+        s.replacen("\\\\?\\", "", 1)
+    } else {
+        s.to_string()
     }
+}
 
-    #[cfg(feature = "webc_runner")]
-    fn generate_pirita_wasmer_main_c(atom_names: &[String], atom_to_run: &str) -> String {
-        let mut c_code_to_add = String::new();
-        let mut c_code_to_instantiate = String::new();
-        let mut deallocate_module = String::new();
+/// Link compiled objects using the system linker
+#[allow(clippy::too_many_arguments)]
+fn link_objects_system_linker(
+    libwasmer_path: &Path,
+    linker_cmd: &str,
+    optimization_flag: &str,
+    object_paths: &[PathBuf],
+    target: &Triple,
+    additional_libraries: &[String],
+    output_path: &Path,
+    debug: bool,
+) -> Result<(), anyhow::Error> {
+    let libwasmer_path = libwasmer_path
+        .canonicalize()
+        .context("Failed to find libwasmer")?;
+    let mut command = Command::new(linker_cmd);
+    let command = command
+        .arg("-Wall")
+        .arg(optimization_flag)
+        .args(object_paths.iter().map(|path| path.canonicalize().unwrap()))
+        .arg(&libwasmer_path)
+        .arg("-target")
+        .arg(format!("{}", target));
+
+    // Add libraries required per platform.
+    // We need userenv, sockets (Ws2_32), advapi32 for some system calls and bcrypt for random numbers.
+    #[cfg(windows)]
+    let command = command
+        .arg("-luserenv")
+        .arg("-lWs2_32")
+        .arg("-ladvapi32")
+        .arg("-lbcrypt");
 
-        for atom_name in atom_names.iter() {
-            let atom_name = Self::normalize_atom_name(atom_name);
-            let atom_name_uppercase = atom_name.to_uppercase();
+    // On unix we need dlopen-related symbols, libmath for a few things, and pthreads.
+    #[cfg(not(windows))]
+    let command = command.arg("-ldl").arg("-lm").arg("-pthread");
```

Comment:
Could you pull this out into a constant at the top of the file so it's easy to tell at a glance which system libraries are required?

```rust
#[cfg(windows)]
const SYSTEM_LIBRARIES: &[&str] = &["userenv", "ws2_32", "advapi32", "bcrypt"];
#[cfg(not(windows))]
const SYSTEM_LIBRARIES: &[&str] = &["dl", "m", "pthread"];

...

for lib in SYSTEM_LIBRARIES {
  command.arg(format!("-l{lib}"));
}
```

Reply:
Done in https://github.com/wasmerio/wasmer/pull/3430/commits/015fc8c34903c2d6de3730f923eb5d10e613f592